### PR TITLE
added tracing

### PR DESCRIPTION
**Pull request by vladshub**

### CODE REVIEW

File: gbus/bus.go [hunk 1]
```diff
@@ -549,17 +551,23 @@ func (b *DefaultBus) monitorAMQPErrors() {
...
-
+	span, ctx := opentracing.StartSpanFromContext(ctx, "sendImpl")
 	defer func() {
 		if err := recover(); err != nil {
 
 			errMsg := fmt.Sprintf("panic recovered panicking err:\n%v\n%s", err, debug.Stack())
 			er = errors.New(errMsg)
+			span.LogFields(slog.Error(er))
 		}
+		b.SenderLock.Unlock()
```
Comment (rhinof) [inline gbus/bus.go]: why do you need the error block to be included in the synced section?

File: gbus/worker.go [hunk 2]
```diff
@@ -219,48 +223,64 @@ func (worker *worker) invokeDeadletterHandler(delivery amqp.Delivery) {
...
-	)
-	if sp != nil {
-		defer sp.Finish()
-	}
-	*/
-	// Update the context with the span for the subsequent reference.
 
 	//handle a message that originated from a deadletter exchange
 	if worker.isDead(delivery) {
+		span.LogFields(slog.Error(errors.New("delivery is dead")))
```
Comment (rhinof) [inline gbus/worker.go]: "delivery is dead" is a bit misleading, think a better trace entry would be "handling deadletter delivery"  or something like that.

File: gbus/worker.go [hunk 3]
```diff
@@ -219,48 +223,64 @@ func (worker *worker) invokeDeadletterHandler(delivery amqp.Delivery) {
...
 		//reject poison message
-		worker.Reject(false, delivery)
+		err = worker.Reject(false, delivery)
+		if err != nil {
+			span.LogFields(slog.Error(err))
+			worker.log().WithError(err).Error("could not reject message")
+		}
 		return
 	}
+	span.LogFields(
```
Comment (rhinof) [inline gbus/worker.go]: the same code is in bus.go lets extract it to a shared function

File: gbus/worker.go [hunk 4]
```diff
@@ -219,48 +223,64 @@ func (worker *worker) invokeDeadletterHandler(delivery amqp.Delivery) {
...
 		worker.invokeDeadletterHandler(delivery)
 		return
 	}
 
 	bm, err := worker.extractBusMessage(delivery)
 	if err != nil {
+		span.LogFields(slog.Error(err), slog.String("grabbit", "message is poison"))
 		//reject poison message
-		worker.Reject(false, delivery)
+		err = worker.Reject(false, delivery)
```
Comment (rhinof) [inline gbus/worker.go]: It's probably worth invoking it with SafeWithRetries

File: gbus/worker.go [hunk 5]
```diff
@@ -269,9 +289,14 @@ func (worker *worker) processMessage(delivery amqp.Delivery, isRPCreply bool) {
 				log.Fields{"message-name": bm.PayloadFQN,
 					"message-type": bm.Semantics}).
 			Warn("Message received but no handlers found")
+		span.LogFields(slog.String("grabbit", "no handlers found"))
 		// worker.log("Message received but no handlers found\nMessage name:%v\nMessage Type:%v\nRejecting message", bm.PayloadFQN, bm.Semantics)
 		//remove the message by acking it and not rejecting it so it will not be routed to a deadletter queue
-		worker.Ack(delivery)
+		err = worker.Ack(delivery)
```
Comment (rhinof) [inline gbus/worker.go]: It's probably worth invoking it with SafeWithRetries

File: gbus/worker.go [hunk 6]
```diff
@@ -281,13 +306,18 @@ func (worker *worker) processMessage(delivery amqp.Delivery, isRPCreply bool) {
 		tx, txErr = worker.txProvider.New()
 		if txErr != nil {
 			worker.log().WithError(txErr).Error("failed to create transaction")
+			span.LogFields(slog.Error(txErr))
 			//reject the message but requeue it so it gets redelivered until we can create transactions
-			worker.Reject(true, delivery)
+			err = worker.Reject(true, delivery)
```
Comment (rhinof) [inline gbus/worker.go]: It's probably worth invoking it with SafeWithRetries

File: gbus/worker.go [hunk 7]
```diff
@@ -303,35 +333,46 @@ func (worker *worker) processMessage(delivery amqp.Delivery, isRPCreply bool) {
...
+					span.LogFields(slog.Error(ackErr))
+					worker.log().WithError(ackErr).Warn("failed to send ack to the broker")
 				}
 
 			} else {
+				span.LogFields(slog.Error(commitErr))
 				worker.log().WithError(commitErr).Error("failed committing transaction")
 				//if the commit failed we will reject the message
-				worker.Reject(false, delivery)
+				err = worker.Reject(false, delivery)
```
Comment (rhinof) [inline gbus/worker.go]: It's probably worth invoking it with SafeWithRetries

File: gbus/bus.go (hunk 1 above)
Comment (vladshub) [reply gbus/bus.go]: It's not an error block, it's part of the defer.

Comment (rhinof) [reply gbus/bus.go]: yes, that is what I meant, why split the b.SenderLock.Lock() and defer b.SenderLock.Unlock() to separate code bloacks ?

File: gbus/worker.go (hunk 4 above)
Comment (vladshub) [reply gbus/worker.go]: worker.Reject already implements safe retries https://github.com/wework/grabbit/blob/fix-opentracing/gbus/worker.go#L192

File: gbus/worker.go (hunk 5 above)
Comment (vladshub) [reply gbus/worker.go]: worker.Reject already implements safe retries https://github.com/wework/grabbit/blob/fix-opentracing/gbus/worker.go#L192

File: gbus/worker.go (hunk 6 above)
Comment (vladshub) [reply gbus/worker.go]: worker.Reject already implements safe retries https://github.com/wework/grabbit/blob/fix-opentracing/gbus/worker.go#L192

File: gbus/worker.go (hunk 7 above)
Comment (vladshub) [reply gbus/worker.go]: worker.Reject already implements safe retries https://github.com/wework/grabbit/blob/fix-opentracing/gbus/worker.go#L192

**Pull request merged**